File: src/frontend/src/api/problems.ts
import { apiRequest } from './http'
import { buildQuery } from '@/utils/query'
import type {
  PageResult,
  ProblemCreatePayload,
  ProblemDetailEx,
  ProblemEditPayload,
  ProblemLanguage,
  ProblemListQuery,
  ProblemSubmissionItem,
  ProblemSummary,
  ProblemTagItem,
  ProblemTestCase,
  Submission,
  TestCaseDraft,
  TestCaseUpsertPayload,
} from '@/types'

export function listProblems(query: ProblemListQuery = {}): Promise<PageResult<ProblemSummary>> {
  return apiRequest('GET', `/problems${buildQuery(query)}`)
}

/** 题目管理视角：该题全员提交列表（创建者与管理角色，docs/contracts/judge.md）；
 * keyword 模糊匹配提交人昵称，language / status / submit_type 精确过滤 */
export function listProblemSubmissions(
  id: string,
  query: {
    page?: number
    page_size?: number
    status?: string
    keyword?: string
    language?: string
    submit_type?: string
  } = {},
): Promise<PageResult<ProblemSubmissionItem>> {
  return apiRequest('GET', `/problems/${id}/submissions${buildQuery(query)}`)
}

/** 题目管理视角：提交详情（统一入口，管理权限 + 归属校验后复用判题装配） */
export function getProblemSubmission(problemId: string, submissionId: string): Promise<Submission> {
  return apiRequest('GET', `/problems/${problemId}/submissions/${submissionId}`)
}

/** 激活标签列表（public：打标选择器与题库筛选） */
export function listActiveTags(): Promise<ProblemTagItem[]> {
  return apiRequest('GET', '/problems/tags')
}
export function getProblem(id: string): Promise<ProblemDetailEx> {
  return apiRequest('GET', `/problems/${id}`)
}
export function createProblem(body: ProblemCreatePayload): Promise<ProblemSummary> {
  return apiRequest('POST', '/problems', body)
}
export function updateProblem(id: string, body: ProblemEditPayload): Promise<ProblemSummary> {
  return apiRequest('PUT', `/problems/${id}`, body)
}
export function replaceTestCases(id: string, cases: TestCaseDraft[]): Promise<null> {
  return apiRequest('PUT', `/problems/${id}/test-cases`, { cases })
}

/** 显式生效：把已通过验题的暂存集晋升为生效集（验题与晋升解耦；点「保存」才生效） */
export function applyTestCases(id: string): Promise<ProblemSummary> {
  return apiRequest('POST', `/problems/${id}/test-cases/apply`)
}

/** 增量更新测试点：只提交变化的行（带 id=修改，input/expected_output 缺省或 null=内容不变、空字符串=清空该侧；无 id=新增；delete_ids=删除）。响应为服务器权威全量列表 */
export function patchTestCases(
  id: string,
  body: { upserts: TestCaseUpsertPayload[]; delete_ids: string[] },
): Promise<{ cases: ProblemTestCase[] }> {
  return apiRequest('PATCH', `/problems/${id}/test-cases`, body)
}

/** 全量替换展示样例（存 problems.samples，不参与判题；≤10 组、单项各 ≤64KB，explanation 选填） */
export function replaceSamples(
  id: string,
  samples: Array<{ input: string; output: string; explanation?: string }>,
): Promise<null> {
  return apiRequest('PUT', `/problems/${id}/samples`, { samples })
}
export function publishProblem(id: string): Promise<ProblemSummary> {
  return apiRequest('POST', `/problems/${id}/publish`)
}
export function archiveProblem(id: string): Promise<ProblemSummary> {
  return apiRequest('POST', `/problems/${id}/archive`)
}
export function initVerification(
  id: string,
  body: { invite_expires_hours?: number },
): Promise<{
  verification_id: string
  invite?: { token: string; expires_at: string | null }
}> {
  return apiRequest('POST', `/problems/${id}/verify`, body)
}
export function resolveVerifyInvite(token: string): Promise<{
  problem_id: string
  problem_title: string
  expires_at: string | null
  background: string
  description: string
  input_description?: string | null
  output_description?: string | null
  note?: string | null
  tags: string[]
  time_limit_ms: number
  memory_limit_mb: number
  samples: Array<{ name: string; input: string; output: string; explanation?: string }>
}> {
  return apiRequest('GET', `/verify-invites/${token}`)
}

/** 查询题目当前有效的验题邀请链接；无或已失效返回 null */
export function getVerifyInvite(
  id: string,
): Promise<{ token: string; expires_at: string | null } | null> {
  return apiRequest('GET', `/problems/${id}/verify/invite`)
}

/** 提交验题代码：存在进行中验题时任意登录用户可提交（invite_token 可选） */
export function submitVerifyCode(
  id: string,
  body: { code: string; language: ProblemLanguage; invite_token?: string },
): Promise<{ submission_id: string; status: string }> {
  return apiRequest('POST', `/problems/${id}/verify`, body)
}

File: src/backend/app/api/v1/problems.py
"""题库路由（docs/contracts/problems.md /problems* 端点，统一前缀 /api/v1）。

验题提交端点 POST /problems/{id}/verify 在 judge 模块路由中注册
（创建判题提交并派发）。
"""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query

from app.api.deps import ProblemServiceDep, SessionDep, TagServiceDep
from app.models.user import User
from app.schemas.problem import (
    ProblemCreate,
    ProblemDetail,
    ProblemQuery,
    ProblemSummary,
    ProblemUpdate,
    SamplesUpdate,
    TagPublic,
    TestCaseListOut,
    TestCasesOut,
    TestCasesPatch,
    TestCasesUpdate,
    VerificationInviteOut,
)
from app.core.dependency import get_current_user, get_optional_user
from app.core.exceptions import AUTH_NOT_LOGGED_IN, PARAM_FORMAT_INVALID, APIError
from app.utils.pagination import PaginatedResponse
from app.utils.response import ApiResponse, ok

router = APIRouter(tags=["problems"])


@router.get("/problems", response_model=ApiResponse[PaginatedResponse[ProblemSummary]])
async def list_problems(
    service: ProblemServiceDep,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    keyword: str | None = Query(default=None, max_length=128),
    tag: str | None = Query(default=None, max_length=64),
    scope: str = Query(default="all"),
    status: str | None = Query(default=None),
    difficulty_min: int | None = Query(default=None, ge=0),
    difficulty_max: int | None = Query(default=None, ge=0),
    user: User | None = Depends(get_optional_user),
) -> ApiResponse[PaginatedResponse[ProblemSummary]]:
    try:
        query = ProblemQuery(
            page=page, page_size=page_size, keyword=keyword, tag=tag, scope=scope, status=status,
            difficulty_min=difficulty_min, difficulty_max=difficulty_max,
        )
    except Exception as exc:  # pydantic 校验失败转 1001 信封
        raise APIError(PARAM_FORMAT_INVALID, "查询参数不合法", 400) from exc
    if query.scope == "mine" and user is None:
        raise APIError(AUTH_NOT_LOGGED_IN, "查看我的题目需要登录", 401)
    rows, total = await service.list_published(query, viewer=user)
    flags = (
        await service.verification_flags([row.id for row in rows]) if query.scope == "mine" else {}
    )
    items: list[ProblemSummary] = []
    for row in rows:
        item = ProblemSummary.model_validate(row)
        item.needs_reverification = flags.get(row.id, False)
        items.append(item)
    await service.attach_counters(items)
    await service.attach_solve_status(items, user)
    return ok(PaginatedResponse(items=items, total=total, page=query.page, page_size=query.page_size))


@router.post("/problems", response_model=ApiResponse[ProblemSummary])
async def create_problem(
    body: ProblemCreate,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[ProblemSummary]:
    problem = await service.create(user, body)
    await db.commit()  # 显式提交：确保数据持久化后再返回（get_db 会再次 commit，但无害）
    summary = ProblemSummary.model_validate(problem)
    await service.attach_counters([summary])
    return ok(summary)


@router.get("/problems/tags", response_model=ApiResponse[list[TagPublic]])
async def list_active_tags(service: TagServiceDep) -> ApiResponse[list[TagPublic]]:
    """激活标签列表（public：打标选择器与题库筛选；docs/contracts/problems.md 端点表）。

    注意必须先于 /problems/{problem_id} 注册，否则 tags 会被当作 uuid 解析。
    """
    rows = await service.list_active()
    return ok([TagPublic.model_validate(row) for row in rows])


@router.get("/problems/{problem_id}", response_model=ApiResponse[ProblemDetail])
async def get_problem(
    problem_id: uuid.UUID,
    service: ProblemServiceDep,
    user: User | None = Depends(get_optional_user),
) -> ApiResponse[ProblemDetail]:
    return ok(await service.get_detail_view(problem_id, user))


@router.put("/problems/{problem_id}", response_model=ApiResponse[ProblemSummary])
async def update_problem(
    problem_id: uuid.UUID, body: ProblemUpdate,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[ProblemSummary]:
    problem = await service.update(user, problem_id, body)
    await db.commit()  # 显式提交：确保数据持久化
    summary = ProblemSummary.model_validate(problem)
    await service.attach_counters([summary])
    return ok(summary)


@router.get("/problems/{problem_id}/test-cases", response_model=ApiResponse[TestCaseListOut])
async def get_test_cases(
    problem_id: uuid.UUID,
    service: ProblemServiceDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[TestCaseListOut]:
    """测试点列表（独立管理端点）：仅题目管理者（admin 或创建者）可读，普通用户 2003。

    详情响应一律不携带测试点；编辑器经本端点回读目标状态（暂存优先）。
    """
    return ok(await service.get_cases_managed(user, problem_id))


@router.put("/problems/{problem_id}/test-cases", response_model=ApiResponse[None])
async def update_test_cases(
    problem_id: uuid.UUID, body: TestCasesUpdate,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[None]:
    """全量替换暂存集（生效集不动，验题通过后晋升；行不可变不物理删除）。"""
    await service.replace_cases(user, problem_id, body)
    await db.commit()  # 显式提交：确保数据持久化
    return ok(None)


@router.patch("/problems/{problem_id}/test-cases", response_model=ApiResponse[TestCasesOut])
async def patch_test_cases(
    problem_id: uuid.UUID, body: TestCasesPatch,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[TestCasesOut]:
    """增量更新暂存集：只提交变化的测试点（带 id=修改 / 无 id=新增 / delete_ids=目标状态移除）。"""
    await service.patch_cases(user, problem_id, body)
    await db.commit()  # 显式提交：确保数据持久化
    problem = await service.problems.get_by_id(problem_id)
    cases = await service.list_cases_view(problem) if problem else []
    return ok(TestCasesOut(cases=cases))


@router.post("/problems/{problem_id}/test-cases/apply", response_model=ApiResponse[ProblemSummary])
async def apply_test_cases(
    problem_id: uuid.UUID,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[ProblemSummary]:
    """显式生效：把已通过验题的暂存集晋升为生效集（验题与晋升解耦）。"""
    problem = await service.apply_pending_cases(user, problem_id)
    await db.commit()
    summary = ProblemSummary.model_validate(problem)
    await service.attach_counters([summary])
    return ok(summary)


@router.put("/problems/{problem_id}/samples", response_model=ApiResponse[None])
async def update_samples(
    problem_id: uuid.UUID, body: SamplesUpdate,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[None]:
    await service.replace_samples(user, problem_id, body)
    await db.commit()  # 显式提交：确保数据持久化
    return ok(None)


@router.post("/problems/{problem_id}/publish", response_model=ApiResponse[ProblemSummary])
async def publish_problem(
    problem_id: uuid.UUID,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[ProblemSummary]:
    problem = await service.publish(user, problem_id)
    await db.commit()  # 显式提交：确保数据持久化
    summary = ProblemSummary.model_validate(problem)
    await service.attach_counters([summary])
    return ok(summary)


@router.post("/problems/{problem_id}/archive", response_model=ApiResponse[ProblemSummary])
async def archive_problem(
    problem_id: uuid.UUID,
    service: ProblemServiceDep,
    db: SessionDep,
    user: User = Depends(get_current_user),
) -> ApiResponse[ProblemSummary]:
    problem = await service.archive(user, problem_id)
    await db.commit()  # 显式提交：确保数据持久化
    summary = ProblemSummary.model_validate(problem)
    await service.attach_counters([summary])
    return ok(summary)


@router.get("/verify-invites/{token}", response_model=ApiResponse[VerificationInviteOut])
async def resolve_verify_invite(token: str, service: ProblemServiceDep) -> ApiResponse[VerificationInviteOut]:
    return ok(await service.resolve_invite(token))
